lib/src/license_detection/token_matcher.dart

```diff
@@ -0,0 +1,375 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+import 'package:pana/src/license_detection/license.dart';
+
+/// Range of tokens in input text that matched to a range of tokens in known license.
+@sealed
+class MatchRange {
+  /// Range of tokens that were found to be a match in input text.
+  Range input;
+
+  /// Range of tokens that were found to be a match in known license.
+  Range source;
+
+  /// Number of tokens that were matched in this range.
+  int tokensClaimed;
```

Comment:
Why not make these `final`?

Then add helper functions 
```dart
offsetMap[offset]!.last.sourceEnd = srcChecksum.end;
        offsetMap[offset]!.last.inputEnd = tgtChecksum.end;
```

Further down in this file when we have:
```dart
      // Check if this source checksum extend the last match
      // and update the last match for this offset accordingly.
      if (offsetMap.containsKey(offset) &&
          (offsetMap[offset]!.last.input.end == tgtChecksum.end - 1)) {
        offsetMap[offset]!.last.sourceEnd = srcChecksum.end;
        offsetMap[offset]!.last.inputEnd = tgtChecksum.end;
        continue;
      }
```
we could instead do:
```dart
      // Check if this source checksum extend the last match
      // and update the last match for this offset accordingly.
      if (offsetMap.containsKey(offset) &&
          (offsetMap[offset]!.last.input.end == tgtChecksum.end - 1)) {
        offsetMap[offset]!.last = offsetMap[offset]!.last.update(
          sourceEnd: srcChecksum.end,
          inputEnd: tgtChecksum.end,
        );
        continue;
      }
```

This could even be simplified into:
```dart
      // Check if this source checksum extend the last match
      // and update the last match for this offset accordingly.
      final matchRanges = offsetMap[offset];
      if (matchRanges != null &&
          (matchRanges.last.input.end == tgtChecksum.end - 1)) {
        matchRanges.last = matchRanges.last.update(
          sourceEnd: srcChecksum.end,
          inputEnd: tgtChecksum.end,
        );
        continue;
      }
```
That way you avoid an unnecessary null-check :D